File: src/api/setting.ts
import { invokeApi } from "./invoke";

export type SettingKey = "AutoStart" | "HideOnLaunch" | "CloseTary";

export interface SettingsSnapshot {
  auto_start: boolean;
  hide_on_launch: boolean;
  close_tary: boolean;
  launch_key: string;
}

export interface AppSettings {
  autoStart: boolean;
  hideOnLaunch: boolean;
  closeToTray: boolean;
  launchKey: string;
}

function mapSettings(settings: SettingsSnapshot): AppSettings {
  return {
    autoStart: settings.auto_start,
    hideOnLaunch: settings.hide_on_launch,
    closeToTray: settings.close_tary,
    launchKey: settings.launch_key,
  };
}

export async function getSettings(): Promise<AppSettings> {
  const settings = await invokeApi<SettingsSnapshot>("get_settings");
  return mapSettings(settings);
}

export async function updateSetting(
  which: SettingKey,
  yes: boolean,
): Promise<void> {
  await invokeApi("update_settings", { which, yes });
}

export async function updateLaunchKey(launchKey: string): Promise<void> {
  await invokeApi("update_launch_key", { launchKey });
}

export async function getAvatarPath(): Promise<string | null> {
  return invokeApi<string | null>("get_avatar_path");
}

File: src-tauri/src/commands/setting.rs
use crate::commands::setting::WhichSetting::{AutoStart, CloseTary, HideOnLaunch};
use crate::models::Settings;
use crate::services::wal::*;
use crate::state::AppState;
use device_query::Keycode;
use serde::{Deserialize, Serialize};
use tauri::State;

#[derive(Serialize, Deserialize, Clone, Copy)]
pub enum WhichSetting {
    AutoStart,
    HideOnLaunch,
    CloseTary,
}

#[tauri::command]
pub fn get_settings(state: State<'_, AppState>) -> Result<Settings, String> {
    let settings = state
        .settings
        .lock()
        .map_err(|_| "failed to lock settings".to_string())?;

    Ok(settings.clone())
}

#[tauri::command]
pub fn get_avatar_path(state: State<'_, AppState>) -> Result<Option<String>, String> {
    let settings = state
        .settings
        .lock()
        .map_err(|_| "failed to lock settings".to_string())?;

    Ok(settings.avatar_path.clone())
}

#[tauri::command]
pub fn update_settings(
    state: State<'_, AppState>,
    yes: bool,
    which: WhichSetting,
) -> Result<(), String> {
    let mut settings = state
        .settings
        .lock()
        .map_err(|_| "failed to lock settings".to_string())?;

    match which {
        AutoStart => {
            update_auto_start_wal(&state, yes)?;
            settings.auto_start = yes;
        }
        HideOnLaunch => {
            update_hide_on_launch_wal(&state, yes)?;
            settings.hide_on_launch = yes;
        }
        CloseTary => {
            update_close_tray_wal(&state, yes)?;
            settings.close_tary = yes;
        }
    }

    Ok(())
}

#[tauri::command]
pub fn update_launch_key(state: State<'_, AppState>, launch_key: String) -> Result<(), String> {
    launch_key
        .parse::<Keycode>()
        .map_err(|_| format!("invalid launch key: {launch_key}"))?;

    let mut settings = state
        .settings
        .lock()
        .map_err(|_| "failed to lock settings".to_string())?;

    update_launch_key_wal(&state, &launch_key)?;
    settings.launch_key = launch_key;

    Ok(())
}
